src/plugins/discover/public/application/components/default_discover_table/default_discover_table.tsx

```diff
@@ -166,12 +172,17 @@ const DefaultDiscoverTableUI = ({
           lazyLoadLastTimeRef.current = time;
           lazyLoadRequestFrameRef.current = requestAnimationFrame(loadMoreRows);
         }
+        // Ensure we have more desired rows in the queue to prevent stalling when we render the
+        // current desired row count
+        if (renderedRowCount + DESIRED_ROWS_LOOKAHEAD > desiredRowCount) {
+          setDesiredRowCount(Math.min(renderedRowCount + DESIRED_ROWS_LOOKAHEAD, rows.length));
+        }
       };
       lazyLoadRequestFrameRef.current = requestAnimationFrame(loadMoreRows);
     }
 
     return () => cancelAnimationFrame(lazyLoadRequestFrameRef.current);
-  }, [showPagination, renderedRowCount, desiredRowCount]);
+  }, [showPagination, renderedRowCount, desiredRowCount, rows.length]);
```

Comment:
do you mind checking if `rows` is also recreated? if so then it might be worth

```typescript
const [rowCount, setRowCount] = useState(rows.length);
```

Reply:
Yes, it's safely set via `setRows` [in the canvas data subscription](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/4776fcd9e0a875ce4b5ac97a73b8952c846afdf2/src/plugins/discover/public/application/view_components/canvas/index.tsx#L89). The changes should propagate safely. I added a unit test for this to be 100% sure, though.